### DISCUSSION TITLE
How to improve DX for looking up the stream id after claiming from an airstream?

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/airdrops#31] [General]
We need to figure out a way to improve the DX in this case:

<img width="500" alt="SCR-20240201-mwwf" src="https://github.com/sablier-labs/v2-periphery/assets/8782666/3caf0bc2-4833-4244-9aff-cc8969be2661">

Ideally, the subgraph should not be required.

Cc @razgraf, @gavriliumircea.

**Comment by razgraf** [comment]
Should we store this in the contracts? As in include an `address => streamId` mapping?

In theory an `index => streamId` one would be safer, having in mind that every address is normally eligible for one claim but we don't enforce that at a smart contract level (only at the backend level).

**Reply by PaulRBerg** [reply]
Sounds like a very good idea.

**Reply by smol-ninja** [reply]
To @razgraf, how does `index => streamId` solve the problem of finding the stream ID for a given user?

**Reply by razgraf** [reply]
You're right, it doesn't solve it, at least not fully onchain. One would still have to access the IPFS file to grab their `index` before querying the contract for a streamId.

The `index` suggestion was a half baked attempt at fixing the possibility of multiple occurrences of the same address in the tree (rendering `address => streamId` incorrect, since it's not enforced at contract level). But maybe we can simply do an `address => streamId[]` mapping instead? Or find another solution..

**Reply by PaulRBerg** [reply]
OK, let's implement the `address => streamId` mapping.

I don't see why we need an array. The UI doesn't allow the same address to appear multiple times in the Merkle tree, does it?

Is it to account for third-party campaigns that do allow this behavior?

**Reply by razgraf** [reply]
Yes, since the behavior is fair-play at the contract level I thought we should take it into consideration, as restricting to a single address could be considered a bug (if not documented).

**Reply by PaulRBerg** [reply]
Right.

Pushing to a mapping with an array will cost more gas than pushing to a mapping with an integer. However, the difference should be small.

Let's go with implementing an `address => streamIds[]` mapping.
